### ISSUE TITLE
Developper documentation relation postgres <--> Ceph

### ISSUE THREAD

**Issue by VannTen** [thoth-station/storages#2658]
It's not very clear from the code or the documentation what Ceph is used for:

I can see [here](https://thoth-station.ninja/docs/developers/storages/index.html#syncing-results-of-a-workflow-run-in-the-cluster) that we store task results,
[here](https://thoth-station.ninja/docs/developers/adviser/developers_guide.html?highlight=ceph#testing-application-against-ceph-and-a-knowledge-graph-database)
how to access it, but there should a high level description of it's role,
comparable to the postgres schema that we can generate with `generate-schema`.

Currently it's a bit hard to get a definite idea of what is inside (and
consequently, what should be deleted, see #2657)

Also, is it only used as an S3 store ? In that case, why reference Ceph
particularly ?

/kind documentation

**Comment by VannTen** [comment]
@harshad16

**Comment by VannTen** [comment]
Let's see if I can summarize what I understand, to see if I really **do**
understand. I'll see if I can work on a PR to add that in docs after
that.

We store documents in Ceph. Those are the results of various kind of
operations.
Those documents are original data (They can't be reconstructed from the db
postgres).
Postgres references those documents.


Follow ups questions:

- I'm not sure to understand what's going on in the sync process. If
  something exists in Ceph, it will be referenced/created in postgres ? 
- Can we map one document in ceph to one entity only in postgres ? In
  other words, do we have a many-to-one relation between ceph documents
  and postgres entries ? (or another relation, or does that depends ?)
- Do the documents in ceph references back to postrgres entries ?
- Which one of them is the single source of truth, postgres or ceph ? Or
  do they each participate in it ?

On Fri, Jul 15, 2022 at 01:12:07AM -0700, Harshad Reddy Nalla wrote:
> Additional answers:
> > Also, is it only used as an S3 store ? In that case, why reference Ceph particularly ?
> 
> Ceph service is an open source storage service, which uses a similar S3 API Call, so in our documentation sometimes we reference s3 call or s3 store. However, Ceph is a service deployed and being used, as it calls are also s3 it would show up in various places.

Yeah, I see what Ceph is. My questions is more, do we use it exclusively
through the S3 API, or do we also use other features, like CephFS or
block storage ?

In the first case, we might drop references to Ceph in docs and in the
code and simply works with an S3 API, which could be backed by any
service providing that S3 API (the fact that it's backed by Ceph would
be an operational detail).

**Comment by harshad16** [comment]
> Let's see if I can summarize what I understand, to see if I really **do** understand. I'll see if I can work on a PR to add that in docs after that. We store documents in Ceph. Those are the results of various kind of operations. Those documents are original data (They can't be reconstructed from the db postgres). Postgres references those documents.

Yes original result data, cant be reconstructed from db postgres, only reconstruction via re-running the operations again.


>  Follow ups questions: - I'm not sure to understand what's going on in the sync process. If something exists in Ceph, it will be referenced/created in postgres ? 

Not true for all sync, some of them are designed in that way for example:
graph-refresh component schedule the package solver which are missing in the postgres db, so it would try to re-sync the document if not in the postgres db from ceph document. 

> - Can we map one document in ceph to one entity only in postgres ? In other words, do we have a many-to-one relation between ceph documents and postgres entries ? (or another relation, or does that depends ?)

The map would be more of many-to-many, 
Postgres has tables and ceph db has multiple directories (should have been different buckets, however, we use one bucket but a different directory in it for different operations). 
for example: adviser result and solver result are two different documents, and they would be used in different places in postgres db tables.

> - Do the documents in ceph references back to postrgres entries ? - Which one of them is the single source of truth, postgres or ceph ? Or do they each participate in it ?

ceph doesnt reference back to postgres, its other way around,
postgres reference ceph data via document-ids

The connotation of both is different, so saying one of it is signle source of truth would  be right.
For the devs of data service, the postgres db would be source of truth as the application is designed on the tables.
for the devs of the data aggregation, the ceph store would be source of truth.
so we have to maintain the sync between both for great results.

> Yeah, I see what Ceph is. My questions is more, do we use it exclusively through the S3 API, or do we also use other features, like CephFS or block storage ? In the first case, we might drop references to Ceph in docs and in the code and simply works with an S3 API, which could be backed by any service providing that S3 API (the fact that it's backed by Ceph would be an operational detail).

we use it through s3 api or [package](https://pypi.org/project/boto3/) with support s3 , don't know what these packages have underlying in their architecture.
please feel free to update the docs, we can discuss the specifics of the docs in the review.

**Comment by mayaCostantini** [comment]
@VannTen I think Harshad provided a great explanation, I don't see any more details to add that could be useful. Thanks @harshad16 !

**Comment by VannTen** [comment]
related: https://github.com/thoth-station/thoth-application/issues/2539

**Comment by goern** [comment]
is this something we can extract/summarize out in to the docs?

**Comment by VannTen** [comment]
I'm not sure.

I think there is two public for the information, Thoth devs and Thoth
ops, and it's not exactly the same information (checkout #2661)

Developer docs should stay in this repo I think, but I could see
operator docs regarding the storage models being centralized with the
rest of the operational documentation (which is the point of the
thoth-application issue if I read it correctly).

**Comment by VannTen** [comment]
/remove-kind bug
/kind documentation
/priority important-longtem

Related (closely) : #2691

**Comment by VannTen** [comment]
/sig stack-guidance
/remove-priority important-soon
also, see #2767
